### PR TITLE
Implement redis event storage

### PR DESCRIPTION
**Pull request by mrbish**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -11,6 +11,9 @@ ENV SUBSCRIPTIONS_PER_NEWSFEED="1024"
 ENV NEWSFEED_ID_LENGTH="128"
 ENV PROCESSOR_CONCURRENCY="4"
 
+ENV REDIS_DSN="redis://redis/0"
+ENV REDIS_CONN_TIMEOUT=5
```
Comment (rmk135) [inline Dockerfile]: It would be great to use less shortcuts so I tend to ask to rename this to `REDIS_CONNECTION_TIMEOUT`.

File: requirements.txt [hunk 2]
```diff
@@ -1,5 +1,6 @@
 dependency-injector>=3.14
 uvloop>=0.14
 aiohttp[speedups]>=3.6
+aioredis==1.3.0
```
Comment (rmk135) [inline requirements.txt]: Let's use `>= 1.3` version. It would be nice to continuously update dependencies for now.

File: Dockerfile [hunk 3]
```diff
@@ -11,6 +11,9 @@ ENV SUBSCRIPTIONS_PER_NEWSFEED="1024"
 ENV NEWSFEED_ID_LENGTH="128"
 ENV PROCESSOR_CONCURRENCY="4"
 
+ENV REDIS_DSN="redis://redis/0"
```
Comment (rmk135) [inline Dockerfile]: Please, don't provide default here. Change to `NOTSET`, please.

File: src/newsfeed/configuration.py [hunk 4]
```diff
@@ -18,6 +18,8 @@ def get_config() -> Dict[str, Any]:
             'event_storage': {
                 'max_newsfeeds': os.getenv('MAX_NEWSFEEDS'),
                 'max_events_per_newsfeed': os.getenv('EVENTS_PER_NEWSFEED'),
+                'redis_dsn': os.getenv('REDIS_DSN', 'redis://redis/0'),
```
Comment (rmk135) [inline src/newsfeed/configuration.py]: Please, don't provide default.

File: src/newsfeed/configuration.py [hunk 5]
```diff
@@ -18,6 +18,8 @@ def get_config() -> Dict[str, Any]:
...
+                'redis_conn_timeout': os.getenv('REDIS_CONN_TIMEOUT', 5),
```
Comment (rmk135) [inline src/newsfeed/configuration.py]: Please, don't provide default.

File: src/newsfeed/infrastructure/event_storages.py [hunk 6]
```diff
@@ -84,6 +88,105 @@ def __init__(self, config: Dict[str, str]):
             del newsfeed_storage[event_index]
 
 
+class RedisEventStorage(EventStorage):
+    """Event storage that stores events in memory."""
```
Comment (rmk135) [inline src/newsfeed/infrastructure/event_storages.py]: Please, update the doc block.

File: src/newsfeed/infrastructure/event_storages.py [hunk 7]
```diff
@@ -84,6 +88,105 @@ def __init__(self, config: Dict[str, str]):
...
+
+    def __init__(self, config: Dict[str, str]):
+        """Initialize queue."""
```
Comment (rmk135) [inline src/newsfeed/infrastructure/event_storages.py]: Please, update the doc block.

File: src/newsfeed/infrastructure/event_storages.py [hunk 8]
```diff
@@ -84,6 +88,105 @@ def __init__(self, config: Dict[str, str]):
...
+        super().__init__(config)
+
+        self._connection = aioredis.pool.ConnectionsPool(
```
Comment (rmk135) [inline src/newsfeed/infrastructure/event_storages.py]: What do you think about just renaming this to `self._pool`?

File: Dockerfile (hunk 3 above)
Comment (rmk135) [inline Dockerfile]: Please, rename this to `EVENT_STORAGE_DSN`.

File: Dockerfile (hunk 1 above)
Comment (rmk135) [inline Dockerfile]: Please, move this to query param of `EVENT_STORAGE_DSN`.

File: docker-compose.yml [hunk 9]
```diff
@@ -7,6 +7,7 @@ services:
     environment:
       LOG_LEVEL: "DEBUG"
       API_BASE_PATH: "/api/"
+      EVENT_STORAGE_DSN: "redis://redis:6379?db=0&connection_timeout=5&minsize=5&maxsize=10"
```
Comment (rmk135) [inline docker-compose.yml]: What do you think if we decrease the number of connections in the pool? I would propose to have 1-4 for development.

File: src/newsfeed/configuration.py [hunk 10]
```diff
@@ -18,8 +18,7 @@ def get_config() -> Dict[str, Any]:
             'event_storage': {
                 'max_newsfeeds': os.getenv('MAX_NEWSFEEDS'),
                 'max_events_per_newsfeed': os.getenv('EVENTS_PER_NEWSFEED'),
-                'redis_dsn': os.getenv('REDIS_DSN', 'redis://redis/0'),
-                'redis_conn_timeout': os.getenv('REDIS_CONN_TIMEOUT', 5),
+                'event_storage_dsn': os.getenv('EVENT_STORAGE_DSN'),
```
Comment (rmk135) [inline src/newsfeed/configuration.py]: It seems we can rename `'event_storage_dsn'` to just `'dsn'` since it's already in `'event_storage'` context. What do you think?

File: src/newsfeed/infrastructure/utils.py [hunk 11]
```diff
@@ -0,0 +1,13 @@
+"""Utils module for infrastructure."""
+
+from urllib.parse import urlparse, parse_qsl
+
+
+def parse_redis_dsn(dsn):
+    """Redis dsn parser."""
+    parsed_dsn = urlparse(dsn)
+    assert parsed_dsn.scheme in ('', 'redis'), ('Unsupported URI scheme', parsed_dsn.scheme)
```
Comment (rmk135) [inline src/newsfeed/infrastructure/utils.py]: Can you describe the case when `scheme == ''`?

File: src/newsfeed/infrastructure/utils.py [hunk 12]
```diff
@@ -0,0 +1,13 @@
+"""Utils module for infrastructure."""
+
+from urllib.parse import urlparse, parse_qsl
+
+
+def parse_redis_dsn(dsn):
```
Comment (rmk135) [inline src/newsfeed/infrastructure/utils.py]: Why did you move this function to `utils` module?

File: src/newsfeed/infrastructure/event_storages.py [hunk 13]
```diff
@@ -179,7 +183,7 @@ def __init__(self, config: Dict[str, str]):
 
     async def delete_by_fqid(self, newsfeed_id: str, event_id: str) -> None:
         """Delete data of specified event."""
-        async with self._connection.get() as connection:
```
Comment (rmk135) [inline src/newsfeed/infrastructure/event_storages.py]: I've realized that we can refactor next two lines

```python
        async with self._pool.get() as connection:
            redis = aioredis.commands.Redis(connection)
````

with an async context manager:

```python
from contextlib import asynccontextmanager

@asynccontextmanager
async def _get_connection():
    async with self._pool.get() as connection:
        redis = aioredis.commands.Redis(connection)
        yield redis
```

In that case we would not need to wrap pool connection into `aioredis.commands.Redis` every time, but just call `self._get_connection()`:

```python
        async with self._get_connection() as redis:
            ...
```

File: docker-compose.yml (hunk 9 above)
Comment (mrbish) [reply docker-compose.yml]: Done

File: src/newsfeed/configuration.py (hunk 10 above)
Comment (mrbish) [reply src/newsfeed/configuration.py]: Done

File: src/newsfeed/infrastructure/utils.py (hunk 11 above)
Comment (mrbish) [reply src/newsfeed/infrastructure/utils.py]: I don't think so :) Fixed.

File: src/newsfeed/infrastructure/utils.py (hunk 12 above)
Comment (mrbish) [reply src/newsfeed/infrastructure/utils.py]: I moved it to a separate module in perspective to reuse in several places and to keep event storage clean.
Which place do you think will be better?

File: src/newsfeed/infrastructure/event_storages.py (hunk 13 above)
Comment (mrbish) [reply src/newsfeed/infrastructure/event_storages.py]: Yeap, it looks cool! Thank you a lot!

File: src/newsfeed/infrastructure/utils.py (hunk 12 above)
Comment (rmk135) [reply src/newsfeed/infrastructure/utils.py]: This place is fine 👍

File: src/newsfeed/infrastructure/event_storages.py [hunk 14]
```diff
@@ -84,6 +92,108 @@ def __init__(self, config: Dict[str, str]):
...
+        redis_config = parse_redis_dsn(config['dsn'])
+        self._pool = aioredis.pool.ConnectionsPool(
+            address=redis_config['address'],
+            db=int(redis_config['db']),
+            create_connection_timeout=int(redis_config['connection_timeout']),
+            minsize=int(redis_config['minsize']),
+            maxsize=int(redis_config['maxsize']),
+            encoding='utf-8',
+        )
+        self._max_newsfeed_ids = int(config['max_newsfeeds'])
```
Comment (rmk135) [inline src/newsfeed/infrastructure/event_storages.py]: We can remove these two params `max_newsfeed_ids` & `max_events_per_newsfeed_id`.

File: src/newsfeed/infrastructure/event_storages.py [hunk 15]
```diff
@@ -84,6 +92,108 @@ def __init__(self, config: Dict[str, str]):
...
+    async def get_by_newsfeed_id(self, newsfeed_id: str) -> Iterable[EventData]:
+        """Get events data from storage."""
+        async with self._get_connection() as redis:
+            newsfeed_storage = []
+            # TODO:
+            #  - Check async
+            #  - Write base serializer
+            #  - Is it correct place for ordering?
+            #  - Ordering by `published_at`?
+            async for key, _ in redis.izscan(f'newsfeed_id:{newsfeed_id}'):
```
Comment (rmk135) [inline src/newsfeed/infrastructure/event_storages.py]: Please, take a look on how we can use lists here so that complexity of that select is O(1).

Comment (mrbish) [reply src/newsfeed/infrastructure/event_storages.py]: If we'll use a list, we'll read the full list with O(n) complexity as well,
